src/tsx/timeline.tsx

```diff
@@ -1,26 +1,165 @@
 import * as React from "react";
+import * as d3 from "d3";
 
-import { INodeCollection, IEmpty } from "./types";
+import { INodeCollection } from "./types";
 import style from "../less/timeline.module.less";
 
-class Timeline extends React.PureComponent<
-  { nodes: INodeCollection; currentNode: string },
-  IEmpty
-> {
+const VIEWBOX_WIDTH = 200;
+const NODE_SPACING = 30;
+const NODE_RADIUS = 5;
+
+const NODES_BEFORE = 3;
+const NODES_AFTER = 3;
+
+interface IProps {
+  nodes: INodeCollection;
+  history: string[];
+}
+
+class Timeline extends React.PureComponent<IProps, { id: string }> {
+  public constructor(props: IProps) {
+    super(props);
+    // Create a random ID, to avoid collisions if we ever have multiple
+    // timelines.
+    this.state = { id: `timeline${Math.round(Math.random() * 10000000)}` };
+  }
+
+  public componentDidMount(): void {
+    this.updateTimeline(this.props.nodes, this.props.history);
+  }
+
+  public componentDidUpdate(): void {
+    this.updateTimeline(this.props.nodes, this.props.history);
+  }
+
+  public updateTimeline(nodes: INodeCollection, history: string[]): void {
+    // Show the recent history
+    const visibleNodes = history
+      .slice(-NODES_BEFORE - 1, -1)
+      .map((id) => ({ id, tense: "past", ...nodes[id] }));
+
+    if (history.length > 0) {
+      // Show the current node
+      const currentId = history[history.length - 1];
+      visibleNodes.push({
+        id: currentId,
+        tense: "present",
+        ...nodes[currentId],
+      });
+      // Show a few nodes into the future
+      for (let step = 0; step < NODES_AFTER; step++) {
+        const nextId = visibleNodes[visibleNodes.length - 1].next;
+        visibleNodes.push({ id: nextId, tense: "future", ...nodes[nextId] });
+      }
+    }
+
+    const nodesWithPosition = visibleNodes.map((node, index) => ({
+      distanceFromStart: Math.max(history.length - NODES_BEFORE - 1, 0) + index,
+      x: VIEWBOX_WIDTH / 2,
+      y: NODE_SPACING / 2 + index * NODE_SPACING,
+      ...node,
+    }));
+    // Add lines between the nodes
+    const lines: {
+      id: string;
+      startX: number;
+      startY: number;
+      endX: number;
+      endY: number;
+    }[] = [];
+    nodesWithPosition.forEach((node) => {
+      const nextNode = nodesWithPosition.find(
+        ({ id, distanceFromStart }) =>
+          id === node.next && distanceFromStart === node.distanceFromStart + 1,
```

Comment:
Do we need to check _id === node.next_, isnt it enough with the second part? (once again thinking about multipathing)

Reply:
See above comment, I'm also thinking about multipathing here :)